### ISSUE TITLE
pkgconfig: Wrong libdir?

### ISSUE THREAD

**Issue by hlandau** [openssl/openssl#23569]
@icing reported a minor issue with the pkgconfig files here: https://github.com/openssl/openssl/discussions/23339#discussioncomment-8452836

>fyi: had some trouble building, since openssl puts the wrong libdir in the pkgconfig files (just the top level dir and not the /lib underneath).

@levitte Any ideas?

**Comment by levitte** [comment]
Er no...  not for the moment.  I'll have a look.

**Comment by levitte** [comment]
I think this is a user error, actually.  Let me ask in the discussion thread

**Comment by icing** [comment]
@levitte I ran

```
./config --prefix=/opt/quic/openssl-master --libdir=/opt/quic/openssl-master/lib
```

and `/opt/quic/openssl-master/lib/pkgconfig/openssl.pc` showed:

```
libdir=/opt/quic/openssl-master
includedir=/opt/quic/openssl-master/include

Name: OpenSSL
Description: Secure Sockets Layer and cryptography libraries and tools
Version: 3.3.0-dev
Requires: libssl libcrypto
```

**Comment by icing** [comment]
fyi: same result when I omitted the `--libdir=`, but we use that regularly due to having seen `lib64` being used on some system...

**Comment by levitte** [comment]
Oh, er, you're using `--libdir` incorrectly.  It shouldn't be a full path, and leaving it empty is a really bad idea

https://github.com/openssl/openssl/blob/09693750601e39574ca7ea335faacfb23bc7282b/INSTALL.md#L338-L350

**Comment by icing** [comment]
Sorry, I expressed myself confusingly. I wanted to say: the same resulting pkgconfig files are generated when I do not specify the libdir arg to configure.

**Comment by levitte** [comment]
Hmmm, that's curious.  I'll double check, but I'm *pretty* sure I don't get that

**Comment by icing** [comment]
@levitte send your machine over to me immediately!

**Comment by levitte** [comment]
Heh 😉 

Awright, I'm getting what I expected (with /lib64 in the libdir paths).  Maybe you're running into some odd path corner case I haven't encountered.  I could try reproduce your circumstances, which is essentially an in-source-tree build.  Do I get it right that you're observing this oddity with the following config command?

```
./config --prefix=/opt/quic/openssl-master
```

**Comment by levitte** [comment]
Still not.

``` console
$ cat /opt/quic/openssl-master/lib64/pkgconfig/libcrypto.pc 
libdir=/opt/quic/openssl-master/lib64
includedir=/opt/quic/openssl-master/include
enginesdir=/opt/quic/openssl-master/lib64/engines-3
modulesdir=/opt/quic/openssl-master/lib64/ossl-modules

Name: OpenSSL-libcrypto
Description: OpenSSL cryptography library
Version: 3.3.0-dev
Libs: -L${libdir} -lcrypto
Libs.private: -ldl -pthread
Cflags: -I${includedir}
$ cat /opt/quic/openssl-master/lib64/pkgconfig/libssl.pc 
libdir=/opt/quic/openssl-master/lib64
includedir=/opt/quic/openssl-master/include

Name: OpenSSL-libssl
Description: Secure Sockets Layer and cryptography libraries
Version: 3.3.0-dev
Requires.private: libcrypto
Libs: -L${libdir} -lssl
Cflags: -I${includedir}
```

The way this is done is that Makefile creates `installdata.pm`, and what's there is then used to generate `exporters/libcrypto.pc` from `exporters/pkg-config/libcrypto.pc.in`.

Here's how it's done in my setup:

``` console 
$ grep mkinstallvars Makefile
builddata.pm: util/mkinstallvars.pl 
	$(PERL) util/mkinstallvars.pl PREFIX=. BINDIR=apps LIBDIR= INCLUDEDIR=include APPLINKDIR=ms ENGINESDIR=engines MODULESDIR=providers "VERSION=$(VERSION)" "LDLIBS=$(LIB_EX_LIBS)" > $@
installdata.pm: util/mkinstallvars.pl 
	$(PERL) util/mkinstallvars.pl "PREFIX=$(INSTALLTOP)" BINDIR=bin "LIBDIR=$(LIBDIR)" INCLUDEDIR=include APPLINKDIR=include/openssl "ENGINESDIR=$(ENGINESDIR)" "MODULESDIR=$(MODULESDIR)" "PKGCONFIGDIR=$(PKGCONFIGDIR)" "CMAKECONFIGDIR=$(CMAKECONFIGDIR)" "LDLIBS=$(LIB_EX_LIBS)" "VERSION=$(VERSION)" > $@
```

The LIBDIR var in those command lines is the variable we're trying to figure out.  In my Makefile, it looks like this:

``` console
$ grep ^LIBDIR= Makefile
LIBDIR=lib64
```

Could you have a look to see what this looks like on your end?

**Comment by icing** [comment]
Ran:

```
>  rm exporters/openssl.pc
> ./config --prefix=/opt/quic/openssl-master
> make install_sw
...
> cat /opt/quic/openssl-master/lib/pkgconfig/libcrypto.pc 
libdir=/opt/quic/openssl-master
includedir=/opt/quic/openssl-master/include
enginesdir=/opt/quic/openssl-master/lib/engines-3
modulesdir=/opt/quic/openssl-master/lib/ossl-modules

Name: OpenSSL-libcrypto
Description: OpenSSL cryptography library
Version: 3.3.0-dev
Libs: -L${libdir} -lcrypto
Libs.private:
Cflags: -I${includedir}
> cat /opt/quic/openssl-master/lib/pkgconfig/libssl.pc
libdir=/opt/quic/openssl-master
includedir=/opt/quic/openssl-master/include

Name: OpenSSL-libssl
Description: Secure Sockets Layer and cryptography libraries
Version: 3.3.0-dev
Requires.private: libcrypto
Libs: -L${libdir} -lssl
Cflags: -I${includedir}
```

Then

```
> grep mkinstallvars Makefile
builddata.pm: util/mkinstallvars.pl
	$(PERL) util/mkinstallvars.pl PREFIX=. BINDIR=apps LIBDIR= INCLUDEDIR=include APPLINKDIR=ms ENGINESDIR=engines MODULESDIR=providers "VERSION=$(VERSION)" "LDLIBS=$(LIB_EX_LIBS)" > $@
installdata.pm: util/mkinstallvars.pl
	$(PERL) util/mkinstallvars.pl "PREFIX=$(INSTALLTOP)" BINDIR=bin "LIBDIR=$(LIBDIR)" INCLUDEDIR=include APPLINKDIR=include/openssl "ENGINESDIR=$(ENGINESDIR)" "MODULESDIR=$(MODULESDIR)" "PKGCONFIGDIR=$(PKGCONFIGDIR)" "CMAKECONFIGDIR=$(CMAKECONFIGDIR)" "LDLIBS=$(LIB_EX_LIBS)" "VERSION=$(VERSION)" > $@
> grep ^LIBDIR= Makefile
LIBDIR=lib
> cat exporters/openssl.pc
libdir=/opt/quic/openssl-master
includedir=/opt/quic/openssl-master/include

Name: OpenSSL
Description: Secure Sockets Layer and cryptography libraries and tools
Version: 3.3.0-dev
Requires: libssl libcrypto
```

**Comment by levitte** [comment]
That's infuriating...

Can you help figure this out?  I'm guessing that something odd is going on in `util/mkinstallvars.pl`.  This patch should help showing what it sees:

``` diff
diff --git a/util/mkinstallvars.pl b/util/mkinstallvars.pl
index 59a432d28c..3ba88b5c38 100644
--- a/util/mkinstallvars.pl
+++ b/util/mkinstallvars.pl
@@ -21,6 +21,7 @@ my @subdirs = qw(BINDIR LIBDIR INCLUDEDIR APPLINKDIR ENGINESDIR MODULESDIR
 
 my %keys = ();
 foreach (@ARGV) {
+    print STDERR "[DEBUG] ARG: $_\n";
     (my $k, my $v) = m|^([^=]*)=(.*)$|;
     $keys{$k} = 1;
     $ENV{$k} = $v;
@@ -47,8 +48,10 @@ foreach my $k (sort keys %keys) {
             $ENV{"${k}_REL"} = $v;
             $v = File::Spec->rel2abs($v, $ENV{PREFIX});
         }
+        printf STDERR "[DEBUG] \$ENV{${k}_REL} = $ENV{'${k}_REL'}\n";
     }
     $ENV{$k} = $v;
+    printf STDERR "[DEBUG] \$ENV{$k} = $ENV{$k}\n";
 }
 
 print <<_____;
```

And then just do this:

``` console
$ rm installdata.pm
$ make exporters/libcrypto.pc
```

Hopefully, the output will tell us something.

It may turn out that this is a bug in perl's File::Spec we've encountered before, where one of rel2abs or abs2rel doesn't work quite right.

Speaking of which, perl version?

**Comment by icing** [comment]
```
> make exporters/libcrypto.pc
perl util/mkinstallvars.pl "PREFIX=/opt/quic/openssl-master" BINDIR=bin "LIBDIR=lib" INCLUDEDIR=include APPLINKDIR=include/openssl "ENGINESDIR=/opt/quic/openssl-master/lib/engines-3" "MODULESDIR=/opt/quic/openssl-master/lib/ossl-modules" "PKGCONFIGDIR=/opt/quic/openssl-master/lib/pkgconfig" "CMAKECONFIGDIR=/opt/quic/openssl-master/lib/cmake/OpenSSL" "LDLIBS= " "VERSION=3.3.0-dev" > installdata.pm
[DEBUG] ARG: PREFIX=/opt/quic/openssl-master
[DEBUG] ARG: BINDIR=bin
[DEBUG] ARG: LIBDIR=lib
[DEBUG] ARG: INCLUDEDIR=include
[DEBUG] ARG: APPLINKDIR=include/openssl
[DEBUG] ARG: ENGINESDIR=/opt/quic/openssl-master/lib/engines-3
[DEBUG] ARG: MODULESDIR=/opt/quic/openssl-master/lib/ossl-modules
[DEBUG] ARG: PKGCONFIGDIR=/opt/quic/openssl-master/lib/pkgconfig
[DEBUG] ARG: CMAKECONFIGDIR=/opt/quic/openssl-master/lib/cmake/OpenSSL
[DEBUG] ARG: LDLIBS=
[DEBUG] ARG: VERSION=3.3.0-dev
[DEBUG] $ENV{APPLINKDIR_REL} =
[DEBUG] $ENV{APPLINKDIR} = /opt/quic/openssl-master/include/openssl
[DEBUG] $ENV{BINDIR_REL} =
[DEBUG] $ENV{BINDIR} = /opt/quic/openssl-master/bin
[DEBUG] $ENV{CMAKECONFIGDIR_REL} =
[DEBUG] $ENV{CMAKECONFIGDIR} = /opt/quic/openssl-master/lib/cmake/OpenSSL
[DEBUG] $ENV{ENGINESDIR_REL} =
[DEBUG] $ENV{ENGINESDIR} = /opt/quic/openssl-master/lib/engines-3
[DEBUG] $ENV{INCLUDEDIR_REL} =
[DEBUG] $ENV{INCLUDEDIR} = /opt/quic/openssl-master/include
[DEBUG] $ENV{LDLIBS} =
[DEBUG] $ENV{LIBDIR_REL} =
[DEBUG] $ENV{LIBDIR} = /opt/quic/openssl-master/lib
[DEBUG] $ENV{MODULESDIR_REL} =
[DEBUG] $ENV{MODULESDIR} = /opt/quic/openssl-master/lib/ossl-modules
[DEBUG] $ENV{PKGCONFIGDIR_REL} =
[DEBUG] $ENV{PKGCONFIGDIR} = /opt/quic/openssl-master/lib/pkgconfig
[DEBUG] $ENV{PREFIX} = /opt/quic/openssl-master
[DEBUG] $ENV{VERSION} = 3.3.0-dev
perl "-I." "-Mconfigdata" "-Minstalldata" "util/dofile.pl" "-oMakefile" exporters/pkg-config/libcrypto.pc.in > exporters/libcrypto.pc
```

Standard macOS /usr/bin/perl:
```
> perl -V
Summary of my perl5 (revision 5 version 30 subversion 3) configuration:
...
```

**Comment by levitte** [comment]
Er...  that looks like what I expected.  Not entirely, the `_REL` variables look strangely empty.  That doesn't matter for .pc files.

Last check to do, show and tell style:

``` console
$ make exporters/openssl.pc
$ grep LIBDIR builddata.pm installdata.pm
builddata.pm:                  $LIBDIR $LIBDIR_REL
builddata.pm:our $LIBDIR             = '/home/levitte/tmp/openssl-master';
builddata.pm:our $LIBDIR_REL         = '.';
installdata.pm:                  $LIBDIR $LIBDIR_REL
installdata.pm:our $LIBDIR             = '/opt/quic/openssl-master/lib64';
installdata.pm:our $LIBDIR_REL         = 'lib64';
$ grep libdir exporters/*.pc
exporters/libcrypto.pc:libdir=/opt/quic/openssl-master/lib64
exporters/libcrypto.pc:Libs: -L${libdir} -lcrypto
exporters/libssl.pc:libdir=/opt/quic/openssl-master/lib64
exporters/libssl.pc:Libs: -L${libdir} -lssl
exporters/openssl.pc:libdir=/opt/quic/openssl-master/lib64
```

**Comment by levitte** [comment]
(note that the libdir in openssl.pc doesn't really matter)

**Comment by gdiscord** [comment]
can I please get some clarification on this. 
Is  it possible to "force" openssl to install to an already existing lib directory?
I've already tried the libdir=lib aproach without any success; at the end of the automated build process the files always get installed to lib64 directory.
Doing the experiment with make build system.

**Comment by Zenju** [comment]
> Oh, er, you're using `--libdir` incorrectly. It shouldn't be a full path, and leaving it empty is a really bad idea
> 
> https://github.com/openssl/openssl/blob/09693750601e39574ca7ea335faacfb23bc7282b/INSTALL.md#L338-L350

FWIW: I had the same issue: the "lib" ending was missing for "libdir" in libcrypto.pc in OpenSSL 3.3 (no such issues with 3.2.1 and earlier). Turns out I was also using full path for --libdir (perhaps it should have been named --libdirname!). But unlike icing, your comment fixed the bug for me! Thanks!

**Comment by levitte** [comment]
> Turns out I was also using full path for `--libdir` (perhaps it should have been named `--libdirname`!)

Perhaps it should be renamed, indeed.  That will be a surprising build break for most, though.  We've had `--libdir` for ages, and it's unfortunate that this name competes with autotools' use of this exact same option, which does cause quite a mess at times...

**Comment by icing** [comment]
Now Homebrew is delivering pkgconfig files with wrong libdirs.

**Comment by nhorman** [comment]
is this perhaps a duplicate of #24298 ?

**Comment by levitte** [comment]
> is this perhaps a duplicate of #24298 ?

It wasn't, FYI.  But considering we couldn't reproduce it, it was difficult to fix it.

**Comment by zeha** [comment]
> > is this perhaps a duplicate of #24298 ?
> It wasn't, FYI. But considering we couldn't reproduce it, it was difficult to fix it.

Does this imply the problem was fixed? If yes, could a reference to the fix be posted?

It would be great if downstreams could fix their openssl packages instead of all the software using openssl starts to apply workarounds.

**Comment by nhorman** [comment]
I think it was fixed in issue #24298 by pull request https://github.com/openssl/openssl/pull/24687

With that change the output of my generated openssl.pc is:
```
prefix=/opt/quic/openssl-master
exec_prefix=${prefix}
libdir=${exec_prefix}/lib64
includedir=${prefix}/include

Name: OpenSSL
Description: Secure Sockets Layer and cryptography libraries and tools
Version: 3.4.0-dev
Requires: libssl libcrypto
```

If you want it pulled in by downstream distros, you'll have to open requests with individual distros to do so

**Comment by satmandu** [comment]
> > Turns out I was also using full path for `--libdir` (perhaps it should have been named `--libdirname`!)
> 
> Perhaps it should be renamed, indeed.  That will be a surprising build break for most, though.  We've had `--libdir` for ages, and it's unfortunate that this name competes with autotools' use of this exact same option, which does cause quite a mess at times...

I'm just now seeing this, as libdir in pkgconfig files broke on 3.3.1 in Chromebrew builds too.

If setting `--libdir` now does not work with a full path set, then that's definitely a change in behavior from before, and the change ought to be documented.

**Comment by levitte** [comment]
I'm reopening this issue, 'cause it appears to have not been fixed.

That being said, I still am unable to reproduce this, and I asked for further feedback, interaction style, [back in February](https://github.com/openssl/openssl/issues/23569#issuecomment-1941758508), and got no response.

To fix this issue, I need active help trying to trace what's going on that causes this issue.

**Comment by levitte** [comment]
> I'll try building with `--libdir` set to `lib` or `lib64` and try rebuilding to see if that fixes things before I open a new issue...

Thank you

**Comment by satmandu** [comment]
> > I'll try building with `--libdir` set to `lib` or `lib64` and try rebuilding to see if that fixes things before I open a new issue...
> 
> Thank you

`i686` and `x86_64` are done, with `armv7l` still building.

Looks like that fixed the issue:

```cat /usr/local/lib64/pkgconfig/openssl.pc
prefix=/usr/local
exec_prefix=${prefix}
libdir=${exec_prefix}/lib64
includedir=${prefix}/include

Name: OpenSSL
Description: Secure Sockets Layer and cryptography libraries and tools
Version: 3.3.2-dev
Requires: libssl libcrypto
```

**Comment by t8m** [comment]
I believe this is fixed now. If not please ping here or open a new issue.
